### ISSUE TITLE
Unittest assertRaises(SystemExit) not recognized

### ISSUE THREAD

**Issue by pewscorner** [microsoft/pylance-release#982]
## Environment data

-   Language Server version: 2021.2.3
-   OS and version: Windows 10
-   Python version (& distribution if applicable, e.g. Anaconda): 3.9.2

## Expected behaviour

The code following ```with self.assertRaises(SystemExit): sys.exit()``` should *not* be grayed out, as the preceding exit is prevented, so the code is in fact executed.

## Actual behaviour

The code following ```with self.assertRaises(SystemExit): sys.exit()``` is grayed out because Pylance thinks it is unreachable. This doesn't happen if python.languageServer is set to Jedi instead of Pylance.

## Logs

```
User belongs to experiment group 'pythonaa'
User belongs to experiment group 'pythonSendEntireLineToREPL'
User belongs to experiment group 'pythonNotDisplayLinterPrompt'
User belongs to experiment group 'ShowPlayIcon - start'
User belongs to experiment group 'DebugAdapterFactory - experiment'
User belongs to experiment group 'PtvsdWheels37 - experiment'
User belongs to experiment group 'UseTerminalToGetActivatedEnvVars - control'
User belongs to experiment group 'LocalZMQKernel - experiment'
User belongs to experiment group 'CollectLSRequestTiming - control'
User belongs to experiment group 'CollectNodeLSRequestTiming - experiment'
User belongs to experiment group 'EnableIPyWidgets - experiment'
User belongs to experiment group 'DeprecatePythonPath - experiment'
User belongs to experiment group 'RunByLine - experiment'
User belongs to experiment group 'CustomEditorSupport - control'
> conda --version
> pyenv root
> python3.7 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python3.6 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python3 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python2 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> py -3.7 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> py -3.6 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> py -3 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> py -2 c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> conda info --json
> python c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
> python c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py -c "import sys;print(sys.executable)"
Starting Pylance language server.
Python interpreter path: D:\Python\Python39\python.exe
Error 2021-02-23 19:55:53: Detection of Python Interpreter for Command python3 and args  failed as file  does not exist
Error 2021-02-23 19:55:53: Failed to get interpreter information for 'D:\Python\Python25\python.exe' [Error: Command failed: "D:\\Python\Python25\python.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\interpreterInfo.py"
Traceback (most recent call last):
  File "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py", line 30, in <module>
    runpy.run_path(module, run_name="__main__")
AttributeError: 'module' object has no attribute 'run_path'

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)] {
  killed: false,
  code: 1,
  signal: null,
  cmd: '"D:\\\\Python\\Python25\\python.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\interpreterInfo.py"'
}
Error 2021-02-23 19:55:53: Failed to get interpreter information for 'D:\Python\Python24\python.exe' [Error: Command failed: "D:\\Python\Python24\python.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\interpreterInfo.py"
Traceback (most recent call last):
  File "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py", line 9, in ?
    import runpy
ImportError: No module named runpy

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)] {
  killed: false,
  code: 1,
  signal: null,
  cmd: '"D:\\\\Python\\Python24\\python.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\interpreterInfo.py"'
}
Error 2021-02-23 19:55:53: Failed to get interpreter information for 'D:\Python\Python26\python.exe' [Error: Command failed: "D:\\Python\Python26\python.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\interpreterInfo.py"
Traceback (most recent call last):
  File "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py", line 30, in <module>
    runpy.run_path(module, run_name="__main__")
AttributeError: 'module' object has no attribute 'run_path'

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:443:11)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)] {
  killed: false,
  code: 1,
  signal: null,
  cmd: '"D:\\\\Python\\Python26\\python.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\interpreterInfo.py"'
}
Error 2021-02-23 19:55:54: Failed to get interpreter information for 'C:\Users\User\AppData\Local\Microsoft\WindowsApps\python.exe' [Error: Command failed: "C:\\Users\User\AppData\Local\Microsoft\WindowsApps\python.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\interpreterInfo.py"
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:443:11)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)] {
  killed: false,
  code: 9009,
  signal: null,
  cmd: '"C:\\\\Users\\User\\AppData\\Local\\Microsoft\\WindowsApps\\python.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\interpreterInfo.py"'
}
Error 2021-02-23 19:55:54: Failed to get interpreter information for 'C:\Users\User\AppData\Local\Microsoft\WindowsApps\python3.exe' [Error: Command failed: "C:\\Users\User\AppData\Local\Microsoft\WindowsApps\python3.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\interpreterInfo.py"
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:443:11)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)] {
  killed: false,
  code: 9009,
  signal: null,
  cmd: '"C:\\\\Users\\User\\AppData\\Local\\Microsoft\\WindowsApps\\python3.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\interpreterInfo.py"'
}
Error 2021-02-23 19:55:54: Failed to get interpreter information for 'C:\Users\User\AppData\Local\Microsoft\WindowsApps\python.exe' [Error: Command failed: "C:\\Users\User\AppData\Local\Microsoft\WindowsApps\python.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\interpreterInfo.py"
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:443:11)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)] {
  killed: false,
  code: 9009,
  signal: null,
  cmd: '"C:\\\\Users\\User\\AppData\\Local\\Microsoft\\WindowsApps\\python.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\interpreterInfo.py"'
}
Error 2021-02-23 19:55:54: Failed to get interpreter information for 'C:\Users\User\AppData\Local\Microsoft\WindowsApps\python3.exe' [Error: Command failed: "C:\\Users\User\AppData\Local\Microsoft\WindowsApps\python3.exe" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\pyvsc-run-isolated.py" "c:\Users\User\.vscode\extensions\ms-python.python-2021.2.582707922\pythonFiles\interpreterInfo.py"
Python was not found; run without arguments to install from the Microsoft Store, or disable this shortcut from Settings > Manage App Execution Aliases.

	at ChildProcess.exithandler (child_process.js:312:12)
	at ChildProcess.emit (events.js:315:20)
	at maybeClose (internal/child_process.js:1021:16)
	at Socket.<anonymous> (internal/child_process.js:443:11)
	at Socket.emit (events.js:315:20)
	at Pipe.<anonymous> (net.js:674:12)] {
  killed: false,
  code: 9009,
  signal: null,
  cmd: '"C:\\\\Users\\User\\AppData\\Local\\Microsoft\\WindowsApps\\python3.exe" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\pyvsc-run-isolated.py" "c:\\Users\\User\\.vscode\\extensions\\ms-python.python-2021.2.582707922\\pythonFiles\\interpreterInfo.py"'
}
> conda --version
> D:\Python\Python39\python.exe -c "
import unittest
loader = unittest.TestLoader()
suites = loader.discover(".", pattern="*unittest.py")
print("start") #Don't remove this line
for suite in suites._tests:
    for cls in suite._tests:
        try:
            for m in cls._tests:
                print(m.id())
        except:
            pass"
cwd: d:\Documents\Python
> D:\Python\Python39\python.exe -c "
import unittest
loader = unittest.TestLoader()
suites = loader.discover(".", pattern="*unittest.py")
print("start") #Don't remove this line
for suite in suites._tests:
    for cls in suite._tests:
        try:
            for m in cls._tests:
                print(m.id())
        except:
            pass"
cwd: d:\Documents\Python
```

## Code Snippet / Additional information

```python
#!/usr/bin/env python3
import unittest
import sys
import contextlib

class Test(unittest.TestCase):
    def test1(self):

        try:
            sys.exit()
        except SystemExit:
            pass
        print('Pylance knows this is reachable')

        with contextlib.suppress(SystemExit):
            sys.exit()
        print('Pylance knows this is also reachable')

        with self.assertRaises(SystemExit):
            sys.exit()
        print('Pylance thinks this is unreachable')

if __name__ == '__main__':
    unittest.main()
```

**Comment by jakebailey** [comment]
The stub annotates `assertRaises` with the requisite bool return type, so I don't know what's going on here.

https://github.com/python/typeshed/blob/16ae4c61201cd8b96b8b22cdfb2ab9e89ba5bcf2/stdlib/unittest/case.pyi#L261-L263

**Comment by erictraut** [comment]
Here's what's going on. @jakebailey, you may recall that until recently we didn't have any support for context managers that swallow exceptions. I was reluctant to add such support because there was no way I could handle all cases in the type checker. I found a way to handle the majority of cases, but this is an example of one that will not work.

The problem is that the type evaluator needs to understand the code flow graph to evaluate symbol types, but context managers conditionally affect the code flow graph based on whether they swallow exceptions or not. To evaluate whether the context manager does or does not swallow exceptions requires the type evaluator to determine its type. So we have a cyclical dependency: need control flow graph -> need to evaluate type -> need control flow graph. The way I broke this cycle is to handle the simple case where the "with" expression is a call to a constructor of a context manager. For this simple form, the `__exit__` method can be looked up to determine if it returns a `bool` or not. This trick covers most cases. In this particular case, however, the "with" expression is not a call to a constructor. Furthermore, it's a call to a function that has multiple overloads, so determining what that function returns requires evaluation of the argument types. 

In short, I don't see a way to make this work.

The only workarounds I can offer:
1. Ignore the fact that the code below the context manager block is "grayed out".
2. Use some other context manager like `contextlib.suppress` to suppress the exception.

**Comment by pewscorner** [comment]
Well, I suppose a quick and dirty workaround - if the gray code drives me crazy - is to wrap the original assert context inside a suppress context like this:
```
        with contextlib.suppress(SystemExit), self.assertRaises(SystemExit):
            sys.exit()
        print('Pylance knows this is reachable')
```
This way the assert still works as expected.

**Comment by erictraut** [comment]
Ah, that's clever!

**Comment by heejaechang** [comment]
closing as by design.

**Comment by erictraut** [comment]
I've found a way to handle this case. It will be addressed in the next release.
